_posts/2021-03-23-attributing-dns-requests-on-windows.md

```diff
@@ -0,0 +1,84 @@
+custom_thumbnail_name:
+category: "Dev Log"
+---
+
+For an application firewall it is very interesting to know which processes query which domains.
+This sounds like an easy thing to achieve, but in practice in turned out to be a rather complicated problem.
+
+When we first looked into how to find this vital piece of information on Windows 10, we found that for some reason, all dns requests are sent from the "DNS-Client" Windows service.
+There were no dns requests _to_ that service on the wire, just from it.
+Upon further investigation we found that Windows applications use the `dnsapi.dll` for taking care of DNS requests.
+These requests are not sent using common DNS packets on the wire, but use a RPC call via an internal IPC mechanism.
+
+This means that we could not attribute DNS requests to processes out of the box.
+
+### The Quick and Dirty Solution
+
+While researching this issue we found that disabling the DNS-Client seemed to be a rather much requested action and quite common for Windows users.
+Most often, Windows users asked for this in order to work around bugs or to improve some aspect of their network performance.
+
+So we decided to take the easy route and just disable the DNS-Client upon installation, because it's only a caching stub resolver with no other functionality, right? Right?
+
+### The Bummer
+
+No.
+
+Apparently, the DNS-Client is much more than "just" and caching stub resolver.
+Maybe it should have been a stronger hint to us that Microsoft removed the ability to stop it directly - we had to change a value in the registry and reboot in order to get rid of it.
+
+The first reports we got, that made us aware that there is much more to this service, were about VPN clients not working correctly.
+Some of these clients wanted to change the configured DNS server in Windows (which is a good thing - it prevent DNS leaking), but as it turned out, the responsible Windows service for handling these changes is ... the DNS-Client!
+With every report of another broken application that we could attribute to the disabled DNS-Client service, we started to realize that we needed a better solution.
+
+All research regarding to finding another way to see the DNS requests without disabling the DNS-Client turned up no viable options - we dismissed the crazy ones:
+- Hijacking the `dnsapi.dll` to force the `DNS_QUERY_WIRE_ONLY` flag - we are not malware.
+- Directly replacing the DNS-Client service - replicating Windows behavior is a nightmare.
+
+### Biting the bullet
+
+With no options available to directly get the needed information, we were left with only one option: Working around the problem on our side.
+
+Until now, we were able to directly attribute a DNS request to a process due to our "quick and dirty" solution.
+No longer being able to do that, does mean that we loose some information, and that we will have a slightly less tight grip on the system.
+We will need to find a good balance between control and configurability.
```

Comment:
Here I got lost since it is hard for me to see a balance between control and configurability. It is later explained what you mean but the reader already is lost. Have to re-work this passage